src/javascript/app/base/subscription_manager.js

```diff
@@ -0,0 +1,204 @@
+import BinarySocket     from './socket';
+import {
+    cloneObject,
+    getPropertyValue,
+    isDeepEqual,
+    isEmptyObject }     from '../../_common/utility';
+
+/**
+ * A layer over BinarySocket to handle subscribing to streaming calls
+ * in order to keep track of subscriptions, manage forget, prevent multiple subscription at the same time, ...
+ *
+ * structure of the the subscription object is:
+ * {
+ *     1: { msg_type: 'proposal', request: { ... }, stream_id: '...', subscribers: [ ... ] },
+ *     2: ...
+ * }
+ * object keys: subscription_id that assigned to each subscription
+ * msg_type   : msg_type of the request for faster filtering
+ * request    : the request object, used to subscribe to the same stream when there is a new subscribe request with exactly the same values
+ * stream_id  : id of the stream which stored from its response and used to forget the stream when needed
+ * subscribers: an array of callbacks to dispatch the response to
+ */
+const SubscriptionManager = (() => {
+    const subscriptions    = {};
+    const forget_requested = {};
+
+    let subscription_id = 0;
+
+    /**
+     * To submit request for a new subscription
+     *
+     * @param {String}   msg_type             msg_type of the request
+     * @param {Object}   request_obj          the whole object of the request to be made
+     * @param {Function} fncCallback          callback function to pass the responses to
+     * @param {Boolean}  should_forget_first  when it's true: forgets the previous subscription, then subscribes after receiving the forget response (if any)
+     */
+    const subscribe = (msg_type, request_obj, fncCallback, should_forget_first = false) => {
+        if (should_forget_first) {
+            forget(msg_type, fncCallback).then(() => {
+                subscribe(msg_type, request_obj, fncCallback);
+            });
+            return;
+        }
+
+        let sub_id = Object.keys(subscriptions).find(id => isDeepEqual(request_obj, subscriptions[id].request));
+
+        if (!sub_id) {
+            sub_id = ++subscription_id;
+
+            subscriptions[sub_id] = {
+                msg_type,
+                request    : cloneObject(request_obj),
+                stream_id  : '',             // stream_id will be updated after receiving the response
+                subscribers: [fncCallback],
+            };
+
+            BinarySocket.send(request_obj, {
+                callback: (response) => dispatch(response, sub_id),
+            });
+        } else if (!hasCallbackFunction(sub_id, fncCallback)) {
+            // there is already an active subscription for the very same request which fncCallback is not subscribed to it yet
+            subscriptions[sub_id].subscribers.push(fncCallback);
+        }
```

Comment:
**Q:** What happens if `sub_id` is false and `else if` is not applicable? should we throw?

Reply:
when there isn't value for `sub_id` (is `false`) the `if` condition becomes `true` so it's not important that the `else if` is applicable or not.